### ISSUE TITLE
Gbm.Report

### ISSUE THREAD

**Issue by SimonDedman** [SimonDedman/gbm.auto#57]
Gbm.report function using markdown to create a word document with the full results section and model interpretation, data units n sources (get via gbm.call, have user fill em in, generate template csv file to look for), bfcheck & hyperparameter choice n options (gbm.call), loop

Compile all papers' reviewers' questions and criticisms into a doc and address them for FI paper but also here as structured questions / headings.
Comparative analysis to GLM GAMs (& PCA?)

**Comment by SimonDedman** [comment]
varint, correlations between variables. Pre & post modelling section.

**Comment by SimonDedman** [comment]
Output as a full pdf (& word doc?) for SM which contains everything (inc link to this methodology as a DOI?),
also pdf/word doc for the MS methods subsection with only the required info.

**Comment by SimonDedman** [comment]
Speak to Hugo Flavio whose done a lovely job on this with actel, e.g. https://cran.r-project.org/web/packages/actel/vignettes/a-0_workspace_requirements.html

**Comment by SimonDedman** [comment]
see https://yihui.org/knitr/
and https://kbroman.org/knitr_knutshell/
and https://github.com/hugomflavio/actel/blob/c899647d61183467d807c6277fa6e61de6066663/R/explore.R#L652

very briefly about the html report formatting: looks like it's just sink to html then cat(paste0( then the various chunks you want, via knitr or r directly?
Hugo Flávio  4 hours ago
So, for some background, the very first actel reports (back when it wasn't even called actel) were pdf files, compiled from LaTeX
Hugo Flávio  4 hours ago
So what I did there was sink the LaTeX code into a file, and then use an R command to compile said file to build a pdf
Hugo Flávio  3 hours ago
when I decided to move to rmarkdown, I kept the mechanics. So, I open the rmd file with line 639, cat the whole content into that file and finally exit it in line 749. (edited) 
Hugo Flávio  3 hours ago
This is all happening inside function printExploreRmd(), which is called by explore() in line 519 of the same code file. (edited) 
Simon Dedman  3 hours ago
cheers. reading more about knitr, I wonder if spin and knit2html would be the preferred route if starting fresh (obviously you already had the working code)
Hugo Flávio  3 hours ago
then rmarkdnown::render() comes in in line 531 to compile the html file, from the rmd we created before (edited) 
Hugo Flávio  3 hours ago
ah, one last note: CRAN is very picky about files being written into the user's directory, so all the writting/compiling is happening in tempdir(), and then only the final html file is brought in, in line 535 (edited) 
Hugo Flávio  3 hours ago
with a final touch of automatically opening the report file with browseURL() :slightly_smiling_face:
Hugo Flávio  3 hours ago
I haven't tried any other route, so I don't know, perhaps there is a better way :slightly_smiling_face:
Hugo Flávio  3 hours ago
I find that having an rmd makes it easy and straight forward to edit things. Just go to the right place in the cat() call and edit as desired.
Hugo Flávio  3 hours ago
oh, also, you can check lines 751 to 837, which contain the code for that index on the left side of the report
Simon Dedman  3 hours ago
nice one, thanks for all this dude, very much appreciated
Hugo Flávio  3 hours ago
you are welcome :slightly_smiling_face: feel free to ping me if you want to chat more about it at some point

**Comment by SimonDedman** [comment]
@BonnieAhr you don't need to predictive map to run loop, it'll generate the summaries of various report outputs regardless.

**Comment by BonnieAhr** [comment]
Under Best Binary BRT column in report.csv add # of trees as well. Actually maybe all of the info for the best brt in one column (like deviance and cv correlation) so I don't have to scroll over and find it and check 100 times that I'm looking at the right on. (Potentially even a flag if under 1000 trees). If not in the output maybe could also include a warning that pops up at the end of the run if the best model has under 1000 trees.

**Comment by SimonDedman** [comment]
How to understand/read the ML evaluation plots, what to look for.
Have all the explanation in a document that lives in gbm.auto somewhere, maybe as a dataframe that people can call with "data"?? Or in the vignettes?

**Comment by SimonDedman** [comment]
[2019.10.25 Lies, Damned Lies, and Accuracy Metrics in Machine Learning.odt](https://github.com/SimonDedman/gbm.auto/files/7392293/2019.10.25.Lies.Damned.Lies.and.Accuracy.Metrics.in.Machine.Learning.odt)
[2021-03-02_Ashley_Jester_Stats_Consultation.odt](https://github.com/SimonDedman/gbm.auto/files/7392304/2021-03-02_Ashley_Jester_Stats_Consultation.odt)
evaluating machine learning models: https://learning.oreilly.com/library/view/evaluating-machine-learning/9781492048756/

**Comment by SimonDedman** [comment]
https://www.statology.org/balanced-accuracy/

**Comment by SimonDedman** [comment]
The correlation between predicted habitat values (averaged to 5° x 5° spatial resolution) and CPUE was calculated using a modified t-test from the R package ‘SpatialPack’ (Osorio et al., 2020) which assumes correlations to be inflated by spatial autocorrelation.

Jon Dale paper

**Comment by SimonDedman** [comment]
https://mlr.mlr-org.com/articles/tutorial/measures.html

**Comment by SimonDedman** [comment]
https://scikit-learn.org/stable/modules/model_evaluation.html
see also https://scikit-learn.org/stable/model_selection.html#model-selection
& generally https://scikit-learn.org/stable/index.html

**Comment by SimonDedman** [comment]
Elith 2008: “Predictive performance should not be estimated on training data, but results are provided in Table 3 to show that BRT overfits the data, regardless of careful model development. [...] Statistics on predictive performance can be estimated from the subsets of data excluded from model fitting”, so: CV Mean Correlation | cv.statistics$correlation.mean = cv.cor ?
Even then: how does this compare to e.g. GAMs?

Abeare thesis: “For each of the fitted models, the pseudo-R2, or D2, was calculated for comparison, where: D2 = 1 – (residual deviance/total deviance)”

"Lastly, the predictive performance of the four statistical modeling techniques was evaluated. Root mean square errors (RMSE) were calculated for predictions of log(CPUEyft) made by the fitted models, using the predictor values of the test dataset. Paired t-tests were then used to test the actual versus predicted log(CPUEyft), whereas F-tests were used to test for differences in the variance of predictions between the modeling techniques."

RMSE: https://www.statology.org/how-to-calculate-rmse-in-r/ Calculated manually, needs a prediction vs observation i.e. manual test/train. Could be calculated internally by gbm/step, per k-fold CV?

**Comment by SimonDedman** [comment]
Hastie, T., R. Tibshirani, and J.H. Friedman, 2001. The Elements of Statistical Learning: Data Mining, Inference, and Prediction. Springer-Verlag, New York 

gbm.step metrics:
self.statistics$correlation = self.cor : runs predict.gbm using gbm object created from all user values inc all data, and predicts it to all data.
cv.statistics$correlation.mean = cv.cor. mean of correlation (cor()) for each of k folds of CV.
Should defo use cv.cor NOT self.cor.

cv.statistics$deviance.mean = cv.dev
= dismo::calc.deviance() for each of k folds
"Function to calculate deviance given two vectors of observed and predicted values. Requires a family argument which is set to binomial by default"
Not very helpful. Number can be >1

Added rmse to gbm.step.sd; could also add loads of other Metrics package metrics.

**Comment by SimonDedman** [comment]
Abeare: D2 = 1 – (residual deviance/total deviance)
cv.statistics$deviance.mean = cv.dev
self.statistics$null = total.deviance
is now:
cv.stats$d.squared
in gbm.step.sd

**Comment by SimonDedman** [comment]
Need to work out how to do these for GLM & GAM still. And re-run the models, e.g. for FI paper.

**Comment by SimonDedman** [comment]
Already added deviance explained relative to null % in report. gbm.auto v1.5.7

**Comment by SimonDedman** [comment]
Report to add: 
- in the BRT combos, round the results:
Training Data Correlation: 0.69761649048316
CV Mean Deviance: 0.36294456588329
CV Deviance SE: 0.0144753751984064
CV D squared: 0.999569543999955
CV Mean Correlation: 0.26195952678098
CV Correlation SE: 0.0509443218258805
CV RMSE: 0.217785350991363
DONE

Add package version used somewhere. Not report, separate file?
> packageVersion("gbm.auto")
[1] ‘2023.3.12’
> packageVersion("dismo")
[1] ‘1.3.10’

Similarly sample sizes for bin & gaus?

Could put both in MLEvalMetricsBin.csv but relies on running Bin AND MLEvaluate==TRUE.
Just add another file?
Best place is in Report.
Just before you save report (L1184), add a final column with
gbm.auto version
dismo version
sample size fam1
{sample size fam2}

could add ZI% under ZI T/F?

DONE

**Comment by SimonDedman** [comment]
If doing predictions, once we have the predictions, also predict to the input data then run ML eval metrics on those results: RMSE & %Dev explained (possible?) for continuous predictions, AUC & TSS for binary.

**Comment by mhpob** [comment]
Just chiming in here with a suggestion: I've found great utility in putting a template rmd/qmd file in `inst` and providing all of the info via `quarto::quarto_render(..., execute_params = ...)` or `rmarkdown::render(..., params = ...)`. @hugomflavio did an excellent job putting it all into a function, it just wound up being tough for me to follow/debug in my own code!

Example boilerplate: https://github.com/mhpob/otndo/tree/main/inst/qmd_template
Example rendering function/call: https://github.com/mhpob/otndo/blob/main/R/make_receiver_push_summary.R#L163-L189

**Comment by SimonDedman** [comment]
https://jmlondon.github.io/pathroutr/articles/akharborseal_demo.html another great josh london example
